src/libexpr/parser-state.hh

```diff
@@ -167,49 +168,63 @@ inline Formals * ParserState::validateFormals(Formals * formals, PosIdx pos, Sym
     return formals;
 }
 
+enum IndentChar {
+    Tab = '\t',
+    Space = ' ',
+};
+
+struct StripState {
+    std::optional<IndentChar> indentChar = std::nullopt;
+    bool start = true;
+    size_t min = -1;
+    size_t cur = 0;
+    bool indent(char c) {
+        if (!start) return true;
+        indentChar = indentChar.value_or(IndentChar(c));
+        bool ok = (indentChar == c);
+        cur += ok;
+        return ok;
+    }
+    void stop() {
+        start = false;
+        min = std::min(min, cur);
+    }
+    void reset() {
+        start = true;
+        cur = 0;
+    }
+};
+
 inline Expr * ParserState::stripIndentation(const PosIdx pos,
     std::vector<std::pair<PosIdx, std::variant<Expr *, StringToken>>> && es)
 {
     if (es.empty()) return new ExprString("");
 
     /* Figure out the minimum indentation.  Note that by design
        whitespace-only final lines are not taken into account.  (So
-       the " " in "\n ''" is ignored, but the " " in "\n foo''" is.) */
-    bool atStartOfLine = true; /* = seen only whitespace in the current line */
-    size_t minIndent = 1000000;
-    size_t curIndent = 0;
+       the " " in "\n ''" is ignored, but the " " in "\n foo''" is not.) */
+    std::optional<IndentChar> indentChar = std::nullopt;
+    StripState state;
     for (auto & [i_pos, i] : es) {
-        auto * str = std::get_if<StringToken>(&i);
-        if (!str || !str->hasIndentation) {
-            /* Anti-quotations and escaped characters end the current start-of-line whitespace. */
-            if (atStartOfLine) {
-                atStartOfLine = false;
-                if (curIndent < minIndent) minIndent = curIndent;
-            }
+        StringToken* x = std::get_if<StringToken>(&i);
+        if (!x || !x->hasIndentation) {
+            state.stop();
             continue;
-        }
-        for (size_t j = 0; j < str->l; ++j) {
-            if (atStartOfLine) {
-                if (str->p[j] == ' ')
-                    curIndent++;
-                else if (str->p[j] == '\n') {
-                    /* Empty line, doesn't influence minimum
-                       indentation. */
-                    curIndent = 0;
-                } else {
-                    atStartOfLine = false;
-                    if (curIndent < minIndent) minIndent = curIndent;
-                }
-            } else if (str->p[j] == '\n') {
-                atStartOfLine = true;
-                curIndent = 0;
+        };
+        std::string_view view = *x;
+        for (auto cur: view) {
+            if (cur == '\n') {
+                state.reset();
+                continue;
             }
+            if (!state.indent(cur)) state.stop();
```

Comment:
I'm not sure about the latest change. All I suggested was simplifying the logic in place to make it more readable without impact on performance characteristics. Adding abstractions might increase the overhead measurably, and although I have no numbers to substantiate what I'm saying, I'd argue such refactorings should always come with benchmarks. Nixpkgs and NixOS are huge code bases that already suffer performance issues.

Reply:
Well, I wouldn't call C++ my area of expertise, as I'm more proficient with Rust, but I don't see why would performance suffer in this case.